### PR TITLE
add more pipeline customisation and permissions

### PR DESCRIPTION
**Pull request by winjer**

### CODE REVIEW

File: pipeline-stack.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
-import type { Account, Instance, SubInstance } from './config';
 import { Stack, StackProps, Stage, StageProps } from 'aws-cdk-lib';
-import * as pipelines from 'aws-cdk-lib/pipelines';
 import * as codebuild from 'aws-cdk-lib/aws-codebuild';
+import * as iam from 'aws-cdk-lib/aws-iam';
+import * as pipelines from 'aws-cdk-lib/pipelines';
 import type { Construct } from 'constructs';
-import * as cdk from 'aws-cdk-lib';
+import type { Account, Instance, SubInstance } from './config';
+import type { CdkManager } from './manager';
```
Comment (plumdog) [inline pipeline-stack.ts:7]: Your editor is trash.

File: pipeline-stack.ts [hunk 2]
```diff
@@ -81,6 +81,10 @@ export class PipelineStack<A> extends Stack {
         return [];
     }
 
+    getOtherPipelineProps(accountConfig: Account, pipelineConfig: Instance<A>, stackProps: StackProps) {
```
Comment (plumdog) [inline pipeline-stack.ts]: This should have return-type `Partial<pipelines.CodePipelineProps>`, I think. I'm sad the linter let you get away without one.

File: pipeline-stack.ts [hunk 3]
```diff
@@ -145,5 +150,17 @@ export class PipelineStack<A> extends Stack {
...
 
         this.pipeline.buildPipeline();
+        this.pipeline.synthProject.addToRolePolicy(
+            new iam.PolicyStatement({
+                effect: iam.Effect.ALLOW,
+                actions: ['sts:AssumeRole'],
+                resources: [
+                    // what about if we are deploying to a different region, or to multiple accounts and regions?
+                    `arn:aws:iam::${accountConfig.accountNumber}:role/cdk-hnb659fds-lookup-role-${accountConfig.accountNumber}-${this.region}`,
+                    `arn:aws:iam::${this.account}:role/cdk-hnb659fds-lookup-role-${this.account}-${this.region}`
```
Comment (plumdog) [inline pipeline-stack.ts]: I think these role ARNs should be farmed out to a `getPipelineSynthProjectAllowedRoleArns` that can be overridden, that by default return these ARNs.

And if it returns `[]`, then don't call `.addToRolePolicy`.

File: pipeline-stack.ts (hunk 1 above)
Comment (winjer) [reply pipeline-stack.ts:7]: it is alphabetical by source library i think.

File: pipeline-stack.ts [hunk 4]
```diff
@@ -145,5 +158,16 @@ export class PipelineStack<A> extends Stack {
         );
 
         this.pipeline.buildPipeline();
+        const allowedRoles = this.getPipelineSynthProjectAllowedRoleArns(accountConfig, pipelineConfig, stackProps);
+        if(allowedRoles) {
```
Comment (plumdog) [inline pipeline-stack.ts]:
```
$ node
Welcome to Node.js v16.17.0.
Type ".help" for more information.
> !![]
true
```
Needs to be `if (allowedRoles.length)`

**Pull request merged**